File: app/src/main/java/com/mathocosta/learnbasic/ShoppingItemFormActivity.java
package com.mathocosta.learnbasic;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputLayout;

public class ShoppingItemFormActivity extends AppCompatActivity {

    private TextInputLayout nameTextField, quantityTextField, descriptionTextField;
    private ShoppingListItem item;
    private int itemID;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_shopping_item_form);

        nameTextField = findViewById(R.id.nameTextField);
        quantityTextField = findViewById(R.id.quantityTextField);
        descriptionTextField = findViewById(R.id.descriptionTextField);

        Intent intent = getIntent();
        item = (ShoppingListItem) intent.getSerializableExtra("ITEM");
        itemID = intent.getIntExtra("ITEM_ID", -1);
        if (item != null) {
            nameTextField.getEditText().setText(item.getName().toString());
            quantityTextField.getEditText().setText(String.valueOf(item.getQuantity()));
            descriptionTextField.getEditText().setText(item.getName().toString());
        }
    }

    public void onSaveButtonPressed(View view) {
        String itemName = nameTextField.getEditText().getText().toString();
        String itemDescription = descriptionTextField.getEditText().getText().toString();

        String itemQuantityValue = quantityTextField.getEditText().getText().toString();
        int itemQuantity;
        if (itemQuantityValue.isEmpty()) {
            itemQuantity = 1;
        } else {
            itemQuantity = Integer.parseInt(itemQuantityValue);
        }

        if (itemName.isEmpty()) {
            nameTextField.setError("Campo obrigatório");
        } else {
            if (item != null) {
                item.setName(itemName);
                item.setDescription(itemDescription);
                item.setQuantity(itemQuantity);
            } else {
                item = new ShoppingListItem(itemName, itemDescription, itemQuantity, false);
            }

            Intent intent = new Intent();
            intent.putExtra("ITEM", item);
            intent.putExtra("ITEM_ID", itemID);
            setResult(RESULT_OK, intent);
            finish();
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            setResult(RESULT_CANCELED);
            finish();

            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/mathocosta/learnbasic/activities/ShoppingItemFormActivity.kt
package com.mathocosta.learnbasic.activities

import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.google.android.material.textfield.TextInputLayout
import com.mathocosta.learnbasic.R
import com.mathocosta.learnbasic.databinding.ActivityShoppingItemFormBinding
import com.mathocosta.learnbasic.models.ShoppingListItem

class ShoppingItemFormActivity : AppCompatActivity() {
    private var item: ShoppingListItem? = null
    private var itemID = 0

    private lateinit var binding: ActivityShoppingItemFormBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityShoppingItemFormBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Prepare toolbar
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)

        prepareEditItem()
    }

    private fun prepareEditItem() {
        item = intent.getSerializableExtra("ITEM") as ShoppingListItem?
        itemID = intent.getIntExtra("ITEM_ID", -1)

        item?.let {
            binding.nameTextField.editText?.setText(it.name)
            binding.quantityTextField.editText?.setText(it.quantity.toString())
            binding.descriptionTextField.editText?.setText(it.name)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            setResult(RESULT_CANCELED)
            finish()
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    fun onSaveButtonPressed(view: View?) {
        val itemName = binding.nameTextField.editText?.text.toString()
        val itemDescription = binding.descriptionTextField.editText?.text.toString()
        val itemQuantityTextValue = binding.quantityTextField.editText?.text.toString()
        val itemQuantity = itemQuantityTextValue.toIntOrNull() ?: 1

        if (itemName.isEmpty()) {
            binding.nameTextField.error = "Campo obrigatório"
        } else {
            if (item != null) {
                item?.name = itemName
                item?.description = itemDescription
                item?.quantity = itemQuantity
            } else {
                item = ShoppingListItem(itemName, itemDescription, itemQuantity, false)
            }

            val intent = Intent().apply {
                putExtra("ITEM", item)
                putExtra("ITEM_ID", itemID)
            }

            setResult(RESULT_OK, intent)
            finish()
        }
    }
}